R/as_polars.R

```diff
@@ -202,8 +202,8 @@ as_polars_df.RPolarsLazyGroupBy = function(x, ...) {
 #' @param rechunk A logical flag (default `TRUE`).
 #' Make sure that all data of each column is in contiguous memory.
 #' @param schema named list of DataTypes, or character vector of column names.
-#' Should be the same length as the number of columns of `x`.
-#' If schema names or types do not match `x`, the columns will be renamed/recast.
+#' Should match the number of columns in `x` and correspond to each column in `x` by position.
+#' If a column in `x` does not match the name or type at the same position, it will be renamed/recast.
```

Comment:
Wow I didn't notice that it would automatically rename / recast but that seems like a really bad behavior:

``` r
library(polars)
df = data.frame(a = 1:3, b = 4:6)
as_polars_df(df, schema = list(b = pl$String, y = pl$Int32))
#> shape: (3, 2)
#> ┌─────┬─────┐
#> │ b   ┆ y   │
#> │ --- ┆ --- │
#> │ str ┆ i32 │
#> ╞═════╪═════╡
#> │ 1   ┆ 4   │
#> │ 2   ┆ 5   │
#> │ 3   ┆ 6   │
#> └─────┴─────┘
```

This doesn't work in py-polars:

```python
test = pl.DataFrame(
    {
        "a": [1, 2, 3],
        "b": [4, 5, 6],
    },
    schema={"b": pl.String, "y": pl.Int32},
)
ValueError: the given column-schema names do not match the data dictionary

pl.DataFrame(
    {
        "a": [1, 2, 3],
        "b": [4, 5, 6],
    },
    schema_overrides={"b": pl.String, "y": pl.Int32},
)

shape: (3, 2)
┌─────┬──────┐
│ a   ┆ b    │
│ --- ┆ ---  │
│ i64 ┆ str  │
╞═════╪══════╡
│ 1   ┆ null │
│ 2   ┆ null │
│ 3   ┆ null │
└─────┴──────┘
```
This wasn't introduced in this PR so I don't think it's a blocker, but it should definitely be fixed

Reply:
Like this:

```python
>>> import polars as pl
>>> import pyarrow as pa
>>> data = pa.table({"a": [1, 2, 3], "b": [4, 5, 6]})
>>> pl.from_arrow(data, schema = [("b", pl.String), ("y", pl.Int32)])
shape: (3, 2)
┌─────┬─────┐
│ b   ┆ y   │
│ --- ┆ --- │
│ str ┆ i32 │
╞═════╪═════╡
│ 1   ┆ 4   │
│ 2   ┆ 5   │
│ 3   ┆ 6   │
└─────┴─────┘
```

Reply:
Hum, so I guess they need to harmonize the behavior upstream between `pl.DataFrame` and `pl.from_arrow` then?

Reply:
It is quite surprising that this behavior is performed, especially when specified with the dict type.

In general, order in dict is not guaranteed, so processing based on order is not a good idea.

Reply:
> Hum, so I guess they need to harmonize the behavior upstream between `pl.DataFrame` and `pl.from_arrow` then?

Not sure about that.
Perhaps the `schema` is intended for Series rather than DataFrame (e.g., `pyarrow.Array` or something has no name), since it is guaranteed that the column names already exist in the `arrow::Table` for `as_polars_df`, so we can simply remove the `schema` argument from here and use only `schema_override`.

Reply:
In my opinion, there is no user demand to change column names in `as_polars_df()`.
So simply removing the `schema` argument and leaving only `schema_override` would be sufficient.

In terms of type change, the `schema` argument is more difficult to use than `schema_override` in that all columns must be specified.